### PR TITLE
[KEYCLOAK-4721] Consider Session Language of Realm Also In ReCaptcha

### PR DESCRIPTION
**Pull request by shimkiv**

### CODE REVIEW

File: services/src/main/java/org/keycloak/authentication/forms/RegistrationRecaptcha.java [hunk 1]
```diff
@@ -224,19 +231,29 @@ public String getHelpText() {
...
         property = new ProviderConfigProperty();
         property.setName(SITE_SECRET);
         property.setLabel("Recaptcha Secret");
         property.setType(ProviderConfigProperty.STRING_TYPE);
         property.setHelpText("Google Recaptcha Secret");
-        configProperties.add(property);
-
+        CONFIG_PROPERTIES.add(property);
+
+        REFINED_LANGUAGE_CODES.add("zh-HK");
```
Comment (stianst) [inline services/src/main/java/org/keycloak/authentication/forms/RegistrationRecaptcha.java]: It's not the responsibility of RegistrationRecaptcha to map the defined locale to a "refined locale". Please remove this and use userLocale.getLanguage() directly.

**Review by stianst — CHANGES_REQUESTED** [review]
Please remove refined language code mapping. It does not belong here.

Also please squash to a single commit.

File: services/src/main/java/org/keycloak/authentication/forms/RegistrationRecaptcha.java (hunk 1 above)
Comment (shimkiv) [reply services/src/main/java/org/keycloak/authentication/forms/RegistrationRecaptcha.java]: Unfortunately if we'll use just ```getLanguage()``` then users will never be able to translate the ```reCAPTCHA``` to these 10 ```refined``` languages.
Please refer to the https://developers.google.com/recaptcha/docs/language

**Comment by stianst** [comment]
The language strings we've defined in themes are:

* ca
* de
* en
* es
* fr
* it
* ja
* lt
* nl
* no
* pt-BR
* ru
* sv
* zh-CN

All these codes are available on https://developers.google.com/recaptcha/docs/language with exactly the same string. In fact in your mapping there are actually a number of mistakes (for exsample en -> en-GB, de -> de-CH).

**Comment by shimkiv** [comment]
> All these codes are available on https://developers.google.com/recaptcha/docs/language with exactly the same string

Yes, if we'll use the ```Locale.getLanguage()``` then this approach will work for the predefined languages only. But it will not help in case if user will introduce new translation like ```Chinese (Hong Kong) | zh-HK``` or ```Portuguese (Portugal) | pt-PT```, etc.
Translations like ```zh-CN``` <-> ```zh-HK``` might be similar but I'm not sure.
And the ```reCAPTCHA``` translation itself is not only about the ```I'm not a robot``` text.

> In fact in your mapping there are actually a number of mistakes (for exsample en -> en-GB, de -> de-CH)

These are not the mistakes since, as you might know, the ```en```, ```en-GB```, ```en-US```, ```en-COUNTRY```, etc. might contain different translation.

**Comment by stianst** [comment]
The English translation is US, not en-GB. So that's incorrect. Google also provides en which is what we use so that works fine.

Your argument about additional translation is completely broken and one good reason why I'm against this mapping inside the ReCaptcha. It's hardcoded and there's absolutely no way it would work for user introduced translations.

If someone adds Portuguese (Portugal) they would use pt-PT to declare it in themes and that's exactly the string Google uses. Same for zh-HK.

**Comment by shimkiv** [comment]
Regarding the ```Locale::getLanguage```:

> If someone adds Portuguese (Portugal) they would use pt-PT to declare it in themes and that's exactly the string Google uses. Same for zh-HK.

You're wrong. If user will add ```Portuguese (Portugal)``` translation and we'll follow the approach with ```Locale::getLanguage``` then Google will load the default ```Portuguese | pt``` translation.

```
Locale userLocale = new Locale("pt", "PT");
System.out.println(userLocale.getLanguage());
```

will result in ```pt```
So if user will select the ```pt-PT``` lang on KC side then Google will load the reCAPTCHA translated in default ```pt``` lang.

Compare the 
https://www.gstatic.com/recaptcha/api2/v1517207450225/recaptcha__pt.js
and
https://www.gstatic.com/recaptcha/api2/v1517207450225/recaptcha__pt_pt.js

and you will notice the difference.

===========================================

Regarding the solution:
- to get rid of the ```REFINED_LANGUAGE_CODES``` list we can just use the transformed ```userLocaleStr``` as it was introduced here. 
Are you good with this approach ?

**Comment by stianst** [comment]
You should just use locale.toLanguageTag(). That'll output:

ca
de
en
es
fr
it
ja
lt
nl
no
pt-BR
ru
sv
zh-CN

For the built-in locales, which is exactly what you want.

**Comment by shimkiv** [comment]
Ah, ok.
Will roll the updated version later today.

Thank you.

**Comment by shimkiv** [comment]
#4962

**Pull request closed without merging**